### PR TITLE
Convert shell.nix to nix flake

### PR DESCRIPTION
**Pull request by thorstenweber83**
## Description
This PR converts shell.nix to [nix flake](https://nixos.wiki/wiki/Flakes).
Using a nix flake over shell.nix brings lots of improvements:

- easy dependency updates using `nix flake update`
- evaluation caching
- garbage collection of built derivations is prevented

I tested on NixOS (x86_64-linux).\
It would be nice if someone could test this on an OS X machine!

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+            # pass through no* attributes so they
+            # can be combined (in any order) like this:
+            # nix develop .#noAvr.noArm.noTeensy
+            passthru =
+              {
+                noAvr = (makeShell { inherit arm teensy; avr = false; });
+                noArm = (makeShell { inherit avr teensy; arm = false; });
+                noTeensy = (makeShell { inherit avr arm; teensy = false; });
+              };
```
Comment (sigprof) [inline flake.nix:97]: Interesting trick; so officially `devShells` should be single-level, but apparently having subpackages this way works.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          };
+      in
+      {
+        # default shell with all options enabled
+        # nix develop
+        devShell = makeShell { };
```
Comment (sigprof) [inline flake.nix:103]: Recent Nix would complain that `devShell` is deprecated in favor of `devShells.<system>.default`, but older Nix versions won't use that `default` shell… Maybe we still need to live with warnings for some time until old Nix versions can be assumed to disappear.

**Review by sigprof — COMMENTED** [review]
Cc: @purcell @astridyu @andresilva

I'm not sure whether this conversion would be acceptable at the moment — flakes are still considered to be experimental, and apparently some people still actively oppose them.

Another potential problem here is the inclusion of `.envrc` — unfortunately, direnv does not support something like `.envrc.local` to override the `.envrc` that you don't exactly like, and maintaining local changes to `.envrc` without committing them would be a major pain. As long as direnv works that way, maybe the polite way to add `.envrc` to the repo would be to add `.envrc.example` instead, and add the actual `.envrc` to `.gitignore`.

> * easy dependency updates using `nix flake update`

Not much different from `( cd util && niv update )` though (the `cd util` part here is nonstandard, because some QMK maintainers did not like having the `nix` subdirectory at the top level; actually the need to have both `flake.lock` and `flake.nix` at the top level could be perceived as clutter too).

> * garbage collection of built derivations is prevented

This is actually provided by direnv, and should work with the existing `shell.nix` too if you put `use nix` into `.envrc`.

**Comment by zvecr** [comment]
> Another potential problem here is the inclusion of .envrc — ... maintaining local changes to .envrc without committing them would be a major pain

This would be the case for the workflow I use. Possibly storing one in `utilt/nix/` and deploying would be my preference.

**Comment by thorstenweber83** [comment]
I understand your concerns about flakes being an experimental feature.
The decision to adopt them is not up to me, but I can make a proposal that I think nix users will benefit from.

I didn't know direnv did add any gc roots and judging by the [table in the direnv wiki](https://github.com/direnv/direnv/wiki/Nix) I would need to install some direnv plugin to archive that.
With flake every nix user (even without direnv) will benefit from not having their pkgsCross packages garbage collected.

The issue with adding an extra file in the repo root could probably be resolved by moving flake.* to /util/nix and sym linking flake.nix to the root.
Would there be objections against adding such a sym link?

I will remove .envrc and maybe add an example in /util/nix.

File: flake.nix (hunk 2 above)
Comment (thorstenweber83) [reply flake.nix:103]: I didn't even know about that. But as long as there are people using older nix versions i don't see a way around using `devShell`.

**Comment by sigprof** [comment]
> I would need to install some direnv plugin to archive that

Sorry, I missed the fact that basic direnv also has some minimal nix support… You need something like [nix-direnv](https://github.com/nix-community/nix-direnv) to add GC roots automatically (Home Manager even has a dedicated option to enable it).

> With flake every nix user (even without direnv) will benefit from not having their pkgsCross packages garbage collected.

Hmm, Nix by itself does not add any GC roots for flakes — you need something else for that (such as nix-direnv).

> The issue with adding an extra file in the repo root could probably be resolved by moving flake.* to /util/nix and sym linking flake.nix to the root.

Looks like this does not really work — Nix does not do anything special with the symlink, so it creates another `flake.lock` at the top level instead of looking for it at the real location of `flake.nix`.  And adding symlinks to the repo could be problematic for Windows users.

What seems to work is placing both `flake.nix` and `flake.lock` in `util/nix/`, then using this in `.envrc`:
```sh
use flake ./util/nix
```
Referring to such flake from outside of the repository would need some weird syntax though — e.g., if this ever gets merged to the upstream repo:
```
github:qmk/qmk_firmware?dir=util%2fnix
```

Also it may be possible to use [flake-compat](https://github.com/edolstra/flake-compat/) to provide a traditional `shell.nix` while using a flake internally.

------

But actually there is another major problem here — https://github.com/NixOS/nix/issues/3121: If the flake is inside a git repository, Nix copies all files from that repository to the Nix store before evaluating anything from that flake. And even when `flake.nix` is in a subdirectory, not at the top level of the git repository, Nix still copies the whole content of the repository (not just the subdirectory with the flake). This would probably be unacceptable for the repository as large as `qmk_firmware`.

(Looks like that copying does not include the content of submodules though: [NixOS/nix#6633](https://www.github.com/NixOS/nix/issues/6633), otherwise the amount of copying would be even larger, but even copying `qmk_firmware` without submodules is bad enough.)

So until some form of lazy copying mentioned in https://github.com/NixOS/nix/issues/3121 is implemented, I don't think that converting the existing `shell.nix` to a flake stored directly in `qmk_firmware` would be acceptable.

It could be possible to store the flake in another repository and refer to it with `use flake github:OWNER/REPO/REV` in `.envrc`, but if you want the environment to change together with the currently checked out `qmk_firmware` revision, the only secure way to do that would be to make `.envrc` a symlink, e.g., to `util/nix/envrc-flake` (because `.envrc` itself must not be in the repo); using `source_env util/nix/envrc-flake` to combine the upstream envrc with some custom configuration could be dangerous, because the file specified in `source_env` is not checked by the security framework.

**Comment by thorstenweber83** [comment]
> Hmm, Nix by itself does not add any GC roots for flakes — you need something else for that (such as nix-direnv).

You're right, i had the impression that nix flakes did that by themselves.
I think i misread the scope of this issue: https://github.com/NixOS/nix/issues/2869

So if there are so many issues/objections, i will be closing this pr.

**Pull request closed without merging**